lib/peer.js

```diff
@@ -1,59 +1,141 @@
- * @param       {Vector} initial - Initialization Vector for this peer.
+ * @param       {Vector} config - Initialization Vector for this peer.
  * @constructor
  */
- function Peer (init) {
-   this['@data'] = init || {};
-   this.clock = 0;
-   this.stack = [];
-   this.known = {};
-   this.init();
- }
+function Peer (config) {
+  this.config = Object.assign({
+    address: '127.0.0.1',
+    port: 7777
+  }, config || {});
 
-// could be looked up by name of parameter in #3
-util.inherits(Peer, require('./vector'));
+  this.stream = new stream.Transform({
+    transform (chunk, encoding, callback) {
+      console.log('TODO: parse as encrypted data:', chunk);
+      callback(null, chunk);
+    }
+  });
 
-Peer.prototype._connect = async function initiate (notify) {
-  var self = this;
+  // this.stream.on('data', this._handler.bind(this));
 
-  self.emit('log', 'peer connecting...');
+  this.key = this.config.key || new Key();
+  this.hex = this.key.public.encodeCompressed('hex');
+  this.id = crypto.createHash('sha256').update(this.hex).digest('hex');
 
-  var commitment = self._serialize(self['@data']);
-  var seed = crypto.randomBytes(KEY_SIZE);
-  
-  var id = seed;
-  var type = '/types/identity'; // path to _lookup()
-  var store = new Stash(); // TODO: eval() ;)
+  this.address = this.config.address;
+  this.port = this.config.port;
 
-  self.emit('log', ['[PEER]', 'emitting (',type,'):', id ]);
-  self.emit('identity', id);
+  this.connections = {};
 
-  self['@id'] = self['@data'];
-  
-  //await store.close();
+  this.clock = 0;
+  this.stack = [];
+  this.known = {};
 
-  return self;
+  this.init();
+}
 
-  //node.on('error', notify);
+util.inherits(Peer, Vector);
 
-}
+Peer.prototype._connect = function connect (address) {
+  let self = this;
+  let parts = address.split(':');
 
-Peer.prototype.send = function message () {
-  
-}
+  if (parts.length !== 2) return console.error('Invalid address:', address);
+
+  self.connections[address] = new net.Socket();
+  self.connections[address].connect(parts[1], parts[0], function () {
+    let id = Buffer.from(self.id);
+    // self.connections[address].write(message.asRaw());
+  });
+};
+
+Peer.prototype._registerPeer = function registerPeer (socket) {
+  let self = this;
+  let id = [socket.remoteAddress, socket.remotePort].join(':');
+
+  // console.log('[PEER]', 'registering peer:', id);
+
+  socket.on('data', function handler (input) {
+    let message = null;
+    let response = null;
+
+    try {
+      message = Message.fromRaw(input);
+    } catch (E) {
+      console.error('[PEER]', 'error parsing transaction:', E);
+    }
+
+    if (!message) return socket.destroy();
+
+    switch (message.type) {
+      default:
+        console.log('[PEER]', `unhandled message type "${message.type}"`);
+        break;
+      case P2P_PING:
+        response = Message.fromVector([P2P_PONG, message.id]);
+        break;
+      case P2P_INSTRUCTION:
```

Comment:
refactor this whole switch to avoid having a huge on('data') callback in the future